File: src/types/list.rs
/** Serene --- Yet an other Lisp
*
* Copyright (c) 2020  Sameer Rahmani <lxsameer@gnu.org>
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 2 of the License.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
use crate::ast::Expr;
use crate::compiler::Compiler;
use crate::types::core::{ExprResult, Expression};

#[derive(Debug, Eq, PartialEq, Clone)]
pub struct List {
    first: Expr,
    rest: Expr,
    length: u64,
}

impl List {
    pub fn new(first: Expr, rest: Expr) -> List {
        // The order of field definition is important here.
        // If we move the `length` after `rest` we're going
        // to break the ownership rule of rust because `rest: rest`
        // is going to move it to the new list and we can not
        // borrow it afterward.
        List {
            length: match &rest {
                Expr::Cons(v) => v.length + 1,
                _ => {
                    if let Expr::Nil = first {
                        0
                    } else {
                        1
                    }
                }
            },
            first,
            rest,
        }
    }
}

impl Expression for List {
    fn eval() {}
    fn code_gen<'ctx>(&self, compiler: &'ctx Compiler) -> ExprResult<'ctx> {
        Err("Not implemented on list".to_string())
    }
}

File: bootstrap/pkg/core/list.go
/*
 Serene --- Yet an other Lisp

Copyright (c) 2020  Sameer Rahmani <lxsameer@gnu.org>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 2 of the License.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

package core

import (
	"fmt"
	"strings"

	"serene-lang.org/bootstrap/pkg/ast"
	"serene-lang.org/bootstrap/pkg/hash"
)

/** WARNING:
This List implementation may look simple and performant but since
we're using a slice here. But in fact it's not memory effecient at
all. We need to rewrite this later to be a immutable and persistent
link list of cons.
*/

type List struct {
	Node
	ExecutionScope
	exprs []IExpr
}

// Implementing IExpr for List ---

func (l *List) GetType() ast.NodeType {
	return ast.List
}

func (l *List) String() string {
	var strs []string
	for _, e := range l.exprs {
		strs = append(strs, e.String())
	}
	return fmt.Sprintf("(%s)", strings.Join(strs, " "))
}

func (l *List) ToDebugStr() string {
	return fmt.Sprintf("%#v", l)
}

// END: IExpr ---

// Implementing ISeq for List ---

func (l *List) First() IExpr {
	if l.Count() == 0 {
		return MakeNil(MakeNodeFromExpr(l))
	}
	return l.exprs[0]
}

func (l *List) Rest() ISeq {
	if l.Count() < 2 {
		return MakeEmptyList(l.Node)
	}

	rest := l.exprs[1:]
	node := l.Node
	if len(rest) > 0 {
		// n won't be nil here but we should check anyway
		n := MakeNodeFromExprs(rest)
		if n == nil {
			panic("'MakeNodeFromExprs' has returned nil for none empty array of exprs")
		}
		node = *n
	}

	return MakeList(node, rest)
}

func (l *List) Hash() uint32 {
	bytes := []byte("TODO")
	return hash.HashOf(append([]byte{byte(ast.List)}, bytes...))
}

// END: ISeq ---

// Implementing ICountable for List ---

func (l *List) Count() int {
	return len(l.exprs)
}

// END: ICountable ---

// Implementing IColl for List ---

func (l *List) ToSlice() []IExpr {
	return l.exprs
}

func (l *List) Cons(e IExpr) IExpr {
	elements := append([]IExpr{e}, l.ToSlice()...)
	node := MakeNodeFromExprs(elements)

	// Since 'elements' is not empty node won't be nil but we should
	// check anyway
	if node == nil {
		node = &l.Node
	}
	return MakeList(*node, elements)
}

// END: IColl ---

func (l *List) AppendToList(e IExpr) *List {
	l.exprs = append(l.exprs, e)
	return l
}

func ListStartsWith(l *List, sym string) bool {
	if l.Count() > 0 {
		firstElem := l.First()
		if firstElem.GetType() == ast.Symbol {
			return firstElem.(*Symbol).GetName() == sym
		}
	}
	return false
}

func MakeList(n Node, elements []IExpr) *List {
	return &List{
		Node:  n,
		exprs: elements,
	}
}

func MakeEmptyList(n Node) *List {
	return &List{
		Node:  n,
		exprs: []IExpr{},
	}
}
